pkg/controller/machinepool/openstackactuator.go

```diff
@@ -139,37 +138,30 @@ func (a *OpenStackActuator) GenerateMachineSets(cd *hivev1.ClusterDeployment, po
 }
 
 // Get the OS image from an existing master machine.
-func getOpenStackOSImage(masterMachine *machinev1beta1.Machine, scheme *runtime.Scheme, logger log.FieldLogger) (string, error) {
-	providerSpec, err := decodeOpenStackMachineProviderSpec(masterMachine.Spec.ProviderSpec.Value, scheme)
-	if err != nil {
+func getOpenStackOSImage(masterMachine *machinev1beta1.Machine, logger log.FieldLogger) (string, error) {
+	var u unstructured.Unstructured
```

Comment:
The OpenStack provider spec go type is in o/api, so you could use the structured type here rather than unstructured

Reply:
Well, that's the whole issue, isn't it? How would I unmarshal it into a structured type if I don't know which apiVersion it came from? I had considered trying one and then the other... but this seemed more generic.

Reply:
When the providerspecs were migrated from their original locations to the new locations, they remained backward compatible. The api version changed, but from a yaml unmarshalling and re-marshalling, it's absolutely compatible. You should be able to just work with the newer types from o/api

Reply:
I'm still missing something though: How do we unmarshal the old-apiVersion Raw into the new-apiVersion go object? That's what the code was attempting to do before, and it was erroring. Or do we do it by casting the Object instead? Or is it because we were using a decoder that had some internal validation for the apiVersion?

Reply:
Don't use the decoder. The old raw in a standard `json.Unmarshal` will unmarshal without error into the new struct. The decoder validates internally by trying to find the correct go type for a registered and known API group and version, but we weren't registering those because they are raw extensions

Reply:
Okay, I think I get it. Using `json.Unmarshal` doesn't try to validate that the go object was defined in the same library that the json blob's `.apiVersion` field represents. (How would it? That's the point of registering the schema.) So all it would be doing is translating fields using the standard json-to-golang rules and blatting in the values.

Side note: It's legal for the schema to not follow those standard field mapping rules. So in theory we could end up with blank fields in the go object that would have been populated if we were using the schema-based decoder. Right?

Side note: Presumably we would want to use non-strict unmarshaling. If (additive, backward-compatible) changes are made to the APIs and there's a lag before we revendor to incorporate those changes, we don't want the unmarshal to puke because it can't match a field in the struct. (In theory the aforementioned nonstandard field name mappings could also trigger same.)

Side note: Um, if I unmarshaled an old-apiVersion object into a new-apiVersion struct, I would end up with the apiVersion in the struct being "wrong" -- i.e. not matching the library I imported the struct from -- true? As long as I'm careful about how I'm using the thing, it wouldn't be a problem, but I can see it being dangerous to start sending it around the ecosystem in that form...

All that said, pathing the raw json as we ended up doing here seems like not a significantly different degree of evil, so I'm thinking to keep it as is. Are you okay with that?